### DISCUSSION TITLE
Smart search doesn't work at all, returns "Failed to Search Smart - Immich Server Error"

### DISCUSSION THREAD

**Discussion by kikendo** [immich-app/immich#10968] [Q&A] [answered]
### The bug

Whenever I try any search, the application gives this error. I have analyzed all the photos in the library for machine learning.

### The OS that Immich Server is running on

Debian Bullseye

### Version of Immich Server

v1.107.2

### Version of Immich Mobile App

1.107.1 build.162

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
.
```

### Your .env content

```Shell
.
```

### Reproduction steps

```bash
Do a smart search like "car"
```

### Relevant log output

```shell
[Nest] 17  - 07/09/2024, 4:00:29 AM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
    at MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:22:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async MachineLearningRepository.encodeText (/usr/src/app/dist/repositories/machine-learning.repository.js:47:26)
    at async SearchService.searchSmart (/usr/src/app/dist/services/search.service.js:97:27)~pavzxtov] Failed to search smart
[Nest] 17  - 07/09/2024, 4:00:29 AM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
    at MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:22:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async MachineLearningRepository.encodeText (/usr/src/app/dist/repositories/machine-learning.repository.js:47:26)
    at async SearchService.searchSmart (/usr/src/app/dist/services/search.service.js:97:27)~pavzxtov] Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
```

### Additional information

I just noticed that the container "immich_machine_learning" doesn't seem to have a port attached to it when I do "docker ps". I am not sure if these problems are related or if they are, how to fix them.

**Comment by mertalev** [comment]
The error means something failed in the machine learning container. The machine learning logs would be helpful to see what’s going wrong. It could be running out of memory, failing to download the model, etc.

**Reply by kikendo** [reply]
Let me see, is the attached file the relevant logs?
[logs.txt](https://github.com/user-attachments/files/16143307/logs.txt)

Also shouldn't that instance have an assigned network port? `docker ps`
 shows no port:
`CONTAINER ID   IMAGE                                                        COMMAND                  CREATED        STATUS                  PORTS                                       NAMES
1dc49bc57a46   ghcr.io/immich-app/immich-machine-learning:release           "tini -- ./start.sh"     20 hours ago   Up 20 hours (healthy)                                               immich_machine_learning
`

**Reply by mertalev** [reply]
It's having issues downloading the model. I think HF is rate-limiting it or something.

You can either stop the container and try later, or you can download [this](https://huggingface.co/immich-app/ViT-B-32__openai) repo manually and copy it into `/cache/clip` (so final folder path should be `/cache/clip/ViT-B-32__openai`).

**Reply by kikendo** [reply]
copy it into the container r on my main system?
I tried again and this is still failing.

**Reply by PGQT** [reply]
@kikendo it should go here:
```
root@immich:/var/lib/docker/volumes/immich_model-cache/_data/clip/ViT-B-32__openai# ls
README.md    models--immich-app--ViT-B-32__openai  visual
config.json  textual
```
but when i downloaded the model from huggingface, i didnt get all of the files that showed up in an old installation of my immich server from a few months ago, so i am just copying the whole _data directory from one to the other. In short, i seem to be missing a big chunk of the necessary files and data structure to support the ai features.

**Reply by kikendo** [reply]
Im about to just give up on the whole smart search thing, it cannot be this complicated to get it to work.

**Reply by alextran1502** [reply]
it should work out of the box unless your server blocks downloading the model somehow. Are you from mainland China by chance?

**Reply by kikendo** [reply]
No i am not.
I did mention before something nobody seemed to think was relevant, but the AI process has no ports attached and I don't know why. I feel that's the reason why it isn't connecting.

**Reply by alextran1502** [reply]
It doesn't have any port because the communication happens using Docker's internal DNS. Do you have the default setup?

**Reply by kikendo** [reply]
Oh, I see. Well, I just have a default setup. I didn't do anything to it, just followed the instructions on the Wiki for installation.

**Reply by alextran1502** [reply]
Do you mind spinning up a new instance to see if you are still facing the issue? What is your server spec?

**Reply by kikendo** [reply]
I'm not sure I can spin up a new instance at the moment, especially with the amount of photos this one has.
The server is running on system with a Xeon E5-2620 v2 @ 2.10GHz and 32GB of RAM. I don't see how that would affect the ability to download stuff though.

**Reply by alextran1502** [reply]
I mean just spinning up to test out the machine learning portion. And when you perform the search, what are the logs on machine-learning container?

**Reply by kikendo** [reply]
I posted the logs above for the machine learning container.

**Comment by goodman-b** [comment]
Similar issue here. Tested on privileged/unprivilaged  Debian 12 and Ubuntu 22 containers on proxmox.
ML throws errors when using OpenVINO.


immich_machine_learning  | [07/24/24 22:08:58] INFO     Attempt #2 to load textual model 'ViT-B-32__openai' to memory                                          
immich_machine_learning  | [07/24/24 22:08:58] INFO     Setting execution providers to ['OpenVINOExecutionProvider', 'CPUExecutionProvider'], in descending order of preference                                         
immich_machine_learning  | 2024-07-24 22:08:58.167388287 [E:onnxruntime:Default, env.cc:254 ThreadMain] pthread_setaffinity_np failed for thread: 158, index: 1, mask: {4, }, error code: 22 error msg: Invalid argument. Specify the number of threads explicitly so the affinity is not set.

immich_machine_learning  | [07/24/24 22:14:54] ERROR    Exception in ASGI application                      
immich_machine_learning  |                                                                                 
immich_machine_learning  |                              ╭─────── Traceback (most recent call last) ───────╮
immich_machine_learning  |                              │ /usr/src/app/main.py:152 in predict             │
immich_machine_learning  |                              │                                                 │
immich_server            | [Nest] 18  - 07/24/2024, 10:14:54 PM   DEBUG [Api:LoggingInterceptor~wvneolal] POST /api/search/smart 200 365.96ms ::ffff:192.168.50.211
immich_server            | [Nest] 18  - 07/24/2024, 10:14:54 PM VERBOSE [Api:LoggingInterceptor~wvneolal] {"page":1,"withExif":true,"isVisible":true,"query":"dog"}
immich_server            | [Nest] 18  - 07/24/2024, 10:14:54 PM   DEBUG [Api:HttpExceptionFilter~wvneolal] HttpException(500) {"message":"Failed to search smart","error":"Internal Server Error","statusCode":500}
immich_server            | [Nest] 18  - 07/24/2024, 10:14:54 PM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error

**Comment by kikendo** [comment]
I'm still getting this error

**Reply by alextran1502** [reply]
@kikendo did you remove the model_cache volume so that the new model get downloaded?

**Reply by powellclan** [reply]
I have the same issue.  I deleted the model_cache volume and re-created the container.  Same result.

**Reply by feasly** [reply]
I have the same issue, where do I delete the model_cache?

**Comment by kikendo** [comment, marked as answer]
I tried this again in the latest version and smart search seems to be working now. Whatever was happening got fixed at some point between then and now

**Reply by camafive** [reply]
Hello, what exactly did you try to solve the problem? I have the same problem...you only updated to the latest version?

**Reply by kikendo** [reply]
I tried almost everything suggested in this thread then I gave up because none of it was working and some of irt was downright inconvenient and annoying.
That was version 1.107.2
I didn't try again until  version 1.113.1, when it worked.

So yeah upgrade your server if you haven't for whatever reason.

**Reply by camafive** [reply]
i was on 1.113.1 and upgraded to 114 and now there is the problem :(

**Reply by kikendo** [reply]
I'm afraid I cannot be of help, and given the help I get, I doubt you'll get much help from here.
Is your whole system upgraded? Have you updated docker?

**Reply by camafive** [reply]
everything is up to date

**Reply by alextran1502** [reply]
@camafive did you change anything in the machine learning settings? And can you get the logs on the server?

**Reply by camafive** [reply]
I've already solved it. I've removed all the containers and installed it again and also ran smart search again. Thanks

But no, i didn't change anything, only upgrade from 113 to 114. It happened to me twice and started from scratch my Raspberry but this time that was my solution

**Comment by ong-andrew** [comment]
I have the same problem

```
$ docker attach immich_server
[Nest] 17  - 09/06/2024, 2:28:14 PM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
    at MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:22:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async MachineLearningRepository.encodeText (/usr/src/app/dist/repositories/machine-learning.repository.js:47:26)
    at async SearchService.searchSmart (/usr/src/app/dist/services/search.service.js:97:27)~ya88l9in] Failed to search smart
[Nest] 17  - 09/06/2024, 2:28:14 PM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
    at MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:22:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async MachineLearningRepository.encodeText (/usr/src/app/dist/repositories/machine-learning.repository.js:47:26)
    at async SearchService.searchSmart (/usr/src/app/dist/services/search.service.js:97:27)~ya88l9in] Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
```

**Comment by ong-andrew** [comment]
@kikendo I think I figure it out. 

Because remote machine learning is enabled, my desktop has to be on in order to use smart search. 

Can you confirm if this is the same on your end?

**Reply by kikendo** [reply]
No. I already said this has been fixed with the latest version.
https://github.com/immich-app/immich/discussions/10968#discussioncomment-10550324

**Comment by FlyingCatTiger** [comment]
I encountered the same situation on synololy NAS.
After installation, it worked fine. But after updating several times, it encountered a prompt of failed to search smart (immich server error). Reinstalling it didn't work either. Installing it on other NAS also didn't work.

**Reply by FlyingCatTiger** [reply]
After upgrading to v1.118.0 , the problem was solved.
But the immich_server log has ERRORs:
```
[Nest] 8  - 10/16/2024, 12:29:42 AM     LOG [Microservices:PersonService] Detected 2 new faces in asset 76b614b3-b691-437a-ac30-850db5da4778
[Nest] 8  - 10/16/2024, 12:31:44 AM     LOG [Microservices:PersonService] Detected 1 new faces in asset 51cb053e-2826-4f93-8d40-06542136a1f7
[Nest] 36  - 10/16/2024, 12:33:31 AM     LOG [Api:EventRepository] Websocket Disconnect: wmq31dkim6_gcEhiAAAF
[Nest] 36  - 10/16/2024, 12:33:32 AM     LOG [Api:EventRepository] Websocket Connect:    K3B-bjz0c9vG8WSsAAAI
[Nest] 8  - 10/16/2024, 12:33:32 AM   ERROR [Microservices:JobService] Unable to run job handler (smartSearch/smart-search): Error: Machine learning request to "http://immich-machine-learning:3003" failed with Error: write EPIPE
[Nest] 8  - 10/16/2024, 12:33:32 AM   ERROR [Microservices:JobService] Error: Machine learning request to "http://immich-machine-learning:3003" failed with Error: write EPIPE
    at /usr/src/app/dist/repositories/machine-learning.repository.js:19:19
    at async MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:18:21)
    at async MachineLearningRepository.encodeImage (/usr/src/app/dist/repositories/machine-learning.repository.js:42:26)
    at async SmartInfoService.handleEncodeClip (/usr/src/app/dist/services/smart-info.service.js:108:27)
    at async /usr/src/app/dist/services/job.service.js:163:36
    at async Worker.processJob (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:394:28)
    at async Worker.retryIfFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:581:24)
[Nest] 8  - 10/16/2024, 12:33:32 AM   ERROR [Microservices:JobService] Object:
{
  "id": "fe207979-e773-4c9d-8643-b66ea46c7aa1"
}

Error: Missing lock for job 87381. retryJob
    at Scripts.finishedErrors (/usr/src/app/node_modules/bullmq/dist/cjs/classes/scripts.js:266:24)
    at Job.moveToFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/job.js:427:32)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async handleFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:379:21)
    at async Worker.retryIfFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:581:24)
[Nest] 8  - 10/16/2024, 12:33:43 AM     LOG [Microservices:PersonService] Detected 1 new faces in asset 14ea3e16-8ac3-45d0-bd1e-185f436f6d99
[Nest] 8  - 10/16/2024, 12:33:44 AM     LOG [Microservices:PersonService] Detected 2 new faces in asset 0dc41f0c-3119-4a48-8ae2-0dc507d0f111
Error: Missing lock for job 87380. retryJob
    at Scripts.finishedErrors (/usr/src/app/node_modules/bullmq/dist/cjs/classes/scripts.js:266:24)
    at Job.moveToFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/job.js:427:32)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async handleFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:379:21)
    at async Worker.retryIfFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:581:24)
``` 





On another NAS with 6G RAM, the problem still exists.  The immich_machine_learning log has Errors
```
[10/15/24 23:57:34] INFO     Initialized request thread pool with 2 threads.    
[10/15/24 23:57:34] INFO     Application startup complete.                      
[10/15/24 23:58:17] INFO     Loading textual model                              
                             'XLM-Roberta-Large-Vit-B-16Plus' to memory         
[10/15/24 23:58:21] INFO     Setting execution providers to                     
                             ['OpenVINOExecutionProvider',                      
                             'CPUExecutionProvider'], in descending order of    
                             preference                                         
[10/16/24 00:02:39] CRITICAL WORKER TIMEOUT (pid:206)                           
[10/16/24 00:03:17] ERROR    Worker (pid:206) was sent SIGKILL! Perhaps out of  
                             memory?                                            
[10/16/24 00:03:25] INFO     Booting worker with pid: 252                       
[10/16/24 00:04:12] INFO     Started server process [252]                       
[10/16/24 00:04:12] INFO     Waiting for application startup.                   
[10/16/24 00:04:12] INFO     Created in-memory cache with unloading after 300s 
```

**Reply by mertalev** [reply]
It seems the server doesn't have enough memory to run that model in OpenVINO. Compiling the model to OpenVINO format uses much more RAM than normal. I think CPU should work.

**Reply by FlyingCatTiger** [reply]
I see. Thank you.

**Comment by scudo005** [comment]
My 2 cents:
It might be that the server is failing to download the models for face recognition and smart search for whatever reason. 
To check this you can start the container with `docker compose up`, so you are able to see the logs in real time.

**Comment by rxvlc** [comment]
> ### The bug
> Whenever I try any search, the application gives this error. I have analyzed all the photos in the library for machine learning.
> 
> ### The OS that Immich Server is running on
> Debian Bullseye
> 
> ### Version of Immich Server
> v1.107.2
> 
> ### Version of Immich Mobile App
> 1.107.1 build.162
> 
> ### Platform with the issue
> * [x]  Server
> * [ ]  Web
> * [ ]  Mobile
> 
> ### Your docker-compose.yml content
> ```yaml
> .
> ```
> 
> ### Your .env content
> ```shell
> .
> ```
> 
> ### Reproduction steps
> ```shell
> Do a smart search like "car"
> ```
> 
> ### Relevant log output
> ```shell
> [Nest] 17  - 07/09/2024, 4:00:29 AM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
>     at MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:22:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async MachineLearningRepository.encodeText (/usr/src/app/dist/repositories/machine-learning.repository.js:47:26)
>     at async SearchService.searchSmart (/usr/src/app/dist/services/search.service.js:97:27)~pavzxtov] Failed to search smart
> [Nest] 17  - 07/09/2024, 4:00:29 AM   ERROR [Api:Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
>     at MachineLearningRepository.predict (/usr/src/app/dist/repositories/machine-learning.repository.js:22:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async MachineLearningRepository.encodeText (/usr/src/app/dist/repositories/machine-learning.repository.js:47:26)
>     at async SearchService.searchSmart (/usr/src/app/dist/services/search.service.js:97:27)~pavzxtov] Error: Machine learning request '{"clip":{"textual":{"modelName":"ViT-B-32__openai"}}}' failed with status 500: Internal Server Error
> ```
> 
> ### Additional information
> I just noticed that the container "immich_machine_learning" doesn't seem to have a port attached to it when I do "docker ps". I am not sure if these problems are related or if they are, how to fix them.

I've been trying diffrent things like pingin and thoose things and my solution was restarting the containers with docker compose down and docker compose up -d

Then it worked
